File: packages/commerce-types-py/src/warp_commerce_types/session.py
"""Session-level coherence — validate a SEQUENCE of agent actions against the
accumulated history, catching violations that only emerge across steps.

A Python port of the TypeScript ``session.ts``, behaviour-identical. The headline
cross-step property is a CUMULATIVE over-refund: three partial refunds of 80
against a 200 order each individually pass (80 <= 200) but sum to 240 > 200. The
point-in-time I-1 check looks at a commitment's current Refunded state, so a naive
``guard_action``-in-a-loop catches none of these. A session accumulates and checks
the pattern.

This is a COMPOSITION over the proven primitives — it does not fork invariant logic:
  - :func:`guard_action` for per-action validation and planning-oracle alternatives;
  - :func:`check_i1_value_conservation` for the cumulative amount check — the session
    probes the SAME canonical I-1 function with the running refund total, so the
    cumulative rule is the point-in-time rule applied to a sum.

A NOTE ON PARTIAL REFUNDS: the schema models a refund as a single terminal
``Refunded`` state carrying one amount — there is no partial-refund state. So the
session tracks partial refunds in its own ledger (a binding-layer accumulation, not
a schema change) and keeps the order in ``Fulfilled`` until it is fully refunded,
at which point it transitions to ``Refunded``.

MULTI-OBJECT COHERENCE (F6): the per-commitment ledger above caps each commitment
against its OWN committed amount. A second, per-TREE ledger (keyed by the tree ROOT
id) caps the SUM of refunds across a parent and its children against the PARENT's
committed amount — so refunds spread over different children (each individually
valid, each child reconciling to the parent via I-6) cannot cumulatively exceed the
parent. This is ADDITIVE to the per-commitment cap and composes the existing
:func:`check_i6_tree_consistency` (structure) + the same I-1 cumulative probe lifted
to the parent. Standalone commitments are never tree members, so single-commitment
behaviour is unchanged.

Scope: TypeScript and Python. Rust / Go ports are roadmap.
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from ._models import Commitment, CommitmentStateRefunded, Money
from .guard import (
    GuardResult,
    GuardViolation,
    ProposedAction,
    TransitionAlternative,
    World,
    check_version,
    guard_action,
)
from .invariants import check_i1_value_conservation, check_i6_tree_consistency
from .money import add, money_equals
from .transitions import _type_of, is_valid_commitment_transition


def _fmt(x: float) -> str:
    """Print whole amounts without a trailing .0 (200, not 200.0)."""
    return str(int(x)) if float(x).is_integer() else str(x)


def _action_key(action: ProposedAction) -> str:
    """The identity of an action for replay detection. An explicit
    ``idempotency_key`` is used when supplied; otherwise a fingerprint is derived
    from commitment + target type + amount (for a Refunded) + actor. Distinct
    operations therefore need distinct keys to be applied separately."""
    if action.idempotency_key is not None:
        return "key:%s" % action.idempotency_key
    parts = [action.commitment, _type_of(action.to), str(action.actor)]
    if _type_of(action.to) == "Refunded":
        amt = action.to["amount"] if isinstance(action.to, dict) else action.to.amount
        a = amt["amount"] if isinstance(amt, dict) else amt.amount
        cur = amt["currency"] if isinstance(amt, dict) else amt.currency
        parts.extend([str(a), cur])
    return "fp:%s" % "|".join(parts)


def _committed_total(c: Commitment) -> Optional[Money]:
    """Sum the Money in a commitment's ``requested`` subject (single currency)."""
    monies = [v.form.money for v in c.subject.requested if v.form.kind == "Money"]
    if not monies:
        return None
    currency = monies[0].currency
    if any(m.currency != currency for m in monies):
        return None  # mixed: a point-in-time I-1 concern, not ours
    total = monies[0]
    for m in monies[1:]:
        total = add(total, m)
    return total


def _is_cumulative_over_refund(order: Commitment, total: float, currency: str) -> bool:
    """Derived from canonical I-1: probe :func:`check_i1_value_conservation` with a
    commitment in ``Refunded(total)`` state — the point-in-time rule applied to the
    sum, not a second copy of it."""
    probe = order.model_copy(
        update={
            "state": CommitmentStateRefunded(amount=Money(amount=total, currency=currency), at=order.created_at),
            "history": [],
        }
    )
    return any(
        v.invariant == "I-1" and "cannot exceed what was captured" in v.description
        for v in check_i1_value_conservation([probe])
    )


def _tree_root_of(world: World, commitment: Commitment) -> Commitment:
    """The root of ``commitment``'s tree in ``world`` — walk ``parent`` pointers up
    while the parent is present in the world. A commitment with no parent (or whose
    parent is not in the world) is its own root. The append-only ``parent`` /
    ``children`` fields already exist on the model, so this is a pure structural
    read — no schema change."""
    by_id = {str(c.id): c for c in world.commitments}
    current = commitment
    seen: set = set()
    while current.parent is not None:
        parent = by_id.get(str(current.parent))
        if parent is None or str(current.id) in seen:
            break
        seen.add(str(current.id))
        current = parent
    return current


def _is_in_tree(commitment: Commitment, root: Commitment) -> bool:
    """A commitment is "in a tree" if it has a parent (is a child) or has children."""
    return str(root.id) != str(commitment.id) or len(commitment.children) > 0


class Session:
    """A stateful sequence validator over an accumulating world."""

    def __init__(self, initial_world: World) -> None:
        self._world = initial_world
        self._ledger: Dict[str, Dict[str, Any]] = {}  # commitment_id -> {total: Money, count: int}
        # Per-TREE cumulative refund ledger, keyed by the tree ROOT id (F6). The
        # per-commitment ``_ledger`` above caps each commitment against its own
        # committed amount; this caps the SUM of refunds across a parent + its children
        # against the parent's committed amount. Standalone commitments are never tree
        # members, so single-commitment behaviour is byte-for-byte unchanged.
        self._tree_ledger: Dict[str, Dict[str, Any]] = {}  # root_id -> {total: Money, count: int}
        # Idempotency: keys of actions ALREADY APPLIED in this session. Per-session,
        # in-memory — durable cross-session idempotency is not provided (see docs).
        self._applied: set = set()

    @property
    def world(self) -> World:
        """The current accumulated world (updated only on accepted actions)."""
        return self._world

    def refunded_so_far(self, commitment_id: str) -> Optional[Money]:
        """The amount refunded so far for a commitment across this session, or None."""
        tally = self._ledger.get(commitment_id)
        return tally["total"] if tally else None

    def propose(self, action: ProposedAction) -> GuardResult:
        """Validate ``action`` against the accumulated world (and the cross-step
        refund ledger), apply it on success, and return the same verdict as
        :func:`guard_action`. On rejection the world is not advanced.

        Ordering: replay check → conflict check → process. A same-key retry is a
        replay (no-op); a different action planned against a stale version is a
        conflict; an unsafe action is an invariant violation."""
        key = _action_key(action)
        # Replay: an already-applied action is a no-op (world not advanced).
        if key in self._applied:
            return GuardResult(ok=True, next=self._world, replay=True)

        # Optimistic-concurrency conflict (distinct from a replay): if the action was
        # planned against a version the commitment has since moved past, reject as a
        # CONFLICT. Checked here too because the partial-refund path does not route
        # through guard_action.
        conflict_target = next((c for c in self._world.commitments if str(c.id) == action.commitment), None)
        if conflict_target is not None:
            conflict = check_version(conflict_target, action.expected_version)
            if conflict is not None:
                return conflict

        if _type_of(action.to) == "Refunded":
            order = next((c for c in self._world.commitments if str(c.id) == action.commitment), None)
            # If the order can't legally reach Refunded from its current state (e.g. a
            # refund before fulfilment, or after a full refund), let guard_action
            # produce the I-2 rejection WITH alternatives.
            if order is None or not is_valid_commitment_transition(order.state, action.to):
                return guard_action(self._world, action)

            committed = _committed_total(order)
            proposed = action.to["amount"] if isinstance(action.to, dict) else action.to.amount
            proposed_amount = proposed["amount"] if isinstance(proposed, dict) else proposed.amount
            proposed_currency = proposed["currency"] if isinstance(proposed, dict) else proposed.currency

            if committed is not None and proposed_currency == committed.currency:
                prior = self._ledger.get(action.commitment)
                prior_amt = prior["total"].amount if prior else 0.0
                prior_count = prior["count"] if prior else 0
                cumulative = prior_amt + proposed_amount

                if _is_cumulative_over_refund(order, cumulative, committed.currency):
                    remaining = max(0.0, committed.amount - prior_amt)
                    return GuardResult(
                        ok=False,
                        violations=[
                            GuardViolation(
                                rule="I-1",
                                message="Cumulative refunds on %s would reach %s %s across %d refund(s), but "
                                "only %s %s was committed — value is not conserved across the session (the "
                                "point-in-time check sees each refund alone)."
                                % (order.id, _fmt(cumulative), committed.currency, prior_count + 1,
                                   _fmt(committed.amount), committed.currency),
                                fix="Refund at most the remaining %s %s (committed %s − already refunded %s)."
                                % (_fmt(remaining), committed.currency, _fmt(committed.amount), _fmt(prior_amt)),
                            )
                        ],
                        alternatives=[
                            TransitionAlternative(
                                to="Refunded",
                                label="refund the commitment",
                                bounded="cumulative refunds must stay within the committed %s %s; %s %s remains refundable"
                                % (_fmt(committed.amount), committed.currency, _fmt(remaining), committed.currency),
                            )
                        ],
                    )

                # Multi-object coherence (F6): if this commitment is part of a tree (a
                # parent with children, or a child), cap the SUM of refunds across the
                # whole tree against the PARENT's committed amount — so refunds spread
                # over different children (each individually valid, each child
                # reconciling via I-6) cannot cumulatively exceed the parent. Composes
                # the existing check_i6_tree_consistency (structure) + the same I-1
                # cumulative probe (lifted to the parent).
                root = _tree_root_of(self._world, order)
                tree_member = _is_in_tree(order, root)
                if tree_member:
                    children = [c for c in self._world.commitments if c.parent is not None and str(c.parent) == str(root.id)]
                    i6 = check_i6_tree_consistency(root, children)
                    if i6:
                        return GuardResult(
                            ok=False,
                            violations=[GuardViolation(rule=v.invariant, message=v.description, fix=v.fix) for v in i6],
                        )
                    tree_committed = _committed_total(root)
                    if tree_committed is not None and proposed_currency == tree_committed.currency:
                        tree_prior = self._tree_ledger.get(str(root.id))
                        tree_prior_amt = tree_prior["total"].amount if tree_prior else 0.0
                        tree_count = tree_prior["count"] if tree_prior else 0
                        tree_cumulative = tree_prior_amt + proposed_amount
                        if _is_cumulative_over_refund(root, tree_cumulative, tree_committed.currency):
                            remaining = max(0.0, tree_committed.amount - tree_prior_amt)
                            return GuardResult(
                                ok=False,
                                violations=[
                                    GuardViolation(
                                        rule="I-1",
                                        message="Cumulative refunds across the commitment tree rooted at %s would "
                                        "reach %s %s across %d refund(s) on the parent and its children, but the "
                                        "parent committed only %s %s — value is not conserved across the tree."
                                        % (root.id, _fmt(tree_cumulative), tree_committed.currency, tree_count + 1,
                                           _fmt(tree_committed.amount), tree_committed.currency),
                                        fix="Refund at most the remaining %s %s across the tree (parent committed %s "
                                        "− already refunded across the tree %s)."
                                        % (_fmt(remaining), tree_committed.currency, _fmt(tree_committed.amount), _fmt(tree_prior_amt)),
                                    )
                                ],
                                alternatives=[
                                    TransitionAlternative(
                                        to="Refunded",
                                        label="refund the commitment",
                                        bounded="cumulative refunds across the tree must stay within the parent's "
                                        "committed %s %s; %s %s remains refundable"
                                        % (_fmt(tree_committed.amount), tree_committed.currency, _fmt(remaining), tree_committed.currency),
                                    )
                                ],
                            )

                # Accepted refund. Record it. Keep the order in Fulfilled for a PARTIAL
                # refund; transition to Refunded only once the refunds reach committed.
                proposed_money = Money(amount=proposed_amount, currency=proposed_currency)
                new_total = add(prior["total"], proposed_money) if prior else proposed_money

                def _record_tree() -> None:
                    if not tree_member:
                        return
                    tp = self._tree_ledger.get(str(root.id))
                    self._tree_ledger[str(root.id)] = {
                        "total": add(tp["total"], proposed_money) if tp else proposed_money,
                        "count": (tp["count"] if tp else 0) + 1,
                    }

                fully_refunded = money_equals(cumulative, committed.amount, committed.currency)
                if fully_refunded:
                    verdict = guard_action(self._world, action)
                    if not verdict.ok or verdict.next is None:
                        return verdict
                    self._world = verdict.next
                    self._ledger[action.commitment] = {"total": new_total, "count": prior_count + 1}
                    _record_tree()
                    self._applied.add(key)
                    return verdict
                self._ledger[action.commitment] = {"total": new_total, "count": prior_count + 1}
                _record_tree()
                self._applied.add(key)
                return GuardResult(ok=True, next=self._world)

        # Non-refund action: pure compose over guard_action.
        verdict = guard_action(self._world, action)
        if verdict.ok and verdict.next is not None:
            self._world = verdict.next
            self._applied.add(key)
        return verdict


def create_session(initial_world: World) -> Session:
    return Session(initial_world)

File: packages/commerce-types/src/session.ts
/**
 * Session-level coherence — validate a SEQUENCE of agent actions against the
 * accumulated history, catching violations that only emerge across steps.
 *
 * {@link guardAction} validates one action against the current world. Some
 * violations are only visible across a sequence — most importantly a CUMULATIVE
 * over-refund: three partial refunds of 80 against a 200 order each individually
 * pass (80 ≤ 200), but they sum to 240 > 200. The point-in-time I-1 check looks
 * at a commitment's *current* Refunded state, so a naive guardAction-in-a-loop
 * catches none of these. A session accumulates and checks the pattern.
 *
 * This is a COMPOSITION over the proven primitives — it does not fork invariant
 * logic:
 *   - {@link guardAction} for per-action validation (transition table = I-2,
 *     point-in-time audit) and for planning-oracle alternatives on rejection;
 *   - {@link checkI1ValueConservation} for the cumulative amount check — the
 *     session probes the SAME canonical I-1 function with the running refund
 *     total, so the cumulative rule is the point-in-time rule, applied to a sum.
 *
 * Scope (honest): the headline cross-step property is cumulative refund
 * conservation. Ordering that lives on a single commitment (a refund before the
 * order was ever captured/fulfilled) is caught because reaching Refunded from a
 * pre-fulfilment state is not a legal transition — guardAction rejects it with
 * the legal alternatives. Cross-object ordering is covered only to the extent
 * the per-action audit (I-4 temporal integrity) already expresses it; properties
 * the data model cannot express are documented as known limits, not faked.
 *
 * A NOTE ON PARTIAL REFUNDS: the schema models a refund as a single terminal
 * `Refunded` state carrying one amount — there is no partial-refund state. So a
 * session tracks partial refunds in its own ledger (a TS-layer accumulation, not
 * a schema change) and keeps the order in `Fulfilled` until it is fully refunded,
 * at which point it transitions to `Refunded`. The cumulative cap is enforced by
 * the session against the order's committed amount.
 *
 * IDEMPOTENCY & REPLAY-SAFETY: agents retry and networks duplicate, so the SAME
 * action applied twice must not double-apply (a retried refund must not refund
 * twice). The session records the identity of each ACCEPTED action — a
 * caller-supplied `idempotencyKey`, or a derived fingerprint (commitment + target
 * type + amount + actor) when none is given — and treats a repeat as a replay: it
 * does not re-apply, does not advance the world, and returns
 * `{ ok: true, next, replay: true }` so the caller learns "already done", not
 * "applied again". Distinct operations therefore need distinct keys (two
 * structurally-identical refunds with no key are deduped by fingerprint). This is
 * also why a session, not a single `guardAction`, is the home for replay safety:
 * it is the layer that accumulates an applied-action record. Scope is
 * **per-session and in-memory only** — durable, cross-session idempotency would
 * need a persistent store and is not provided here (a documented limit, not a
 * guarantee the session can make).
 *
 * OPTIMISTIC CONCURRENCY: when two actors act on the same commitment, an action
 * may be individually valid yet planned against a STALE version (the other actor
 * advanced the commitment first). A caller passes the version they planned against
 * (`expectedVersion`, from {@link commitmentVersion}, derived from the commitment's
 * append-only history + state — not a schema field); if it no longer matches, the
 * action is rejected as a CONFLICT (`{ ok: false, conflict: true, expected, actual }`)
 * so the caller re-reads and re-plans. A conflict is distinct from an invariant
 * violation (unsafe) and from a replay (a retry, which dedups). This is OPTIMISTIC
 * concurrency over the caller's view — NOT a lock, consensus, or distributed
 * transaction manager; Warp does not serialize concurrent writers.
 *
 * TypeScript first. Ports to Python / Rust / Go are roadmap.
 */

import { checkVersion, guardAction, type GuardResult, type ProposedAction, type World } from "./guard.js";
import { checkI1ValueConservation, checkI6TreeConsistency } from "./invariants.js";
import { add, moneyEquals } from "./money.js";
import type { Money } from "./money.js";
import type { Commitment } from "./primitives.js";
import { isValidCommitmentTransition } from "./transitions.js";

/** The total refunded so far for one commitment, with how many refunds composed it. */
interface RefundTally {
  total: Money;
  count: number;
}

/** A stateful sequence validator over an accumulating world. */
export interface Session {
  /**
   * Validate a proposed action against the ACCUMULATED world (and the session's
   * cross-step ledger), apply it on success, and return the same discriminated
   * verdict as {@link guardAction}. On rejection the world is not advanced.
   */
  propose(action: ProposedAction): GuardResult;
  /** The current accumulated world (read-only view; updated only on accepted actions). */
  readonly world: World;
  /** The amount refunded so far for a commitment across this session, or null if none. */
  refundedSoFar(commitmentId: string): Money | null;
}

/** Sum the Money in a commitment's `requested` subject (single currency). */
function committedTotal(c: Commitment): Money | null {
  const monies: Money[] = [];
  for (const v of c.subject.requested) {
    if (v.form.kind === "Money") monies.push(v.form.money);
  }
  if (monies.length === 0) return null;
  const currency = monies[0]?.currency;
  if (currency === undefined) return null;
  // A mixed-currency subject is a point-in-time I-1 violation already; not ours.
  if (monies.some((m) => m.currency !== currency)) return null;
  return monies.reduce((acc, m) => add(acc, m));
}

/**
 * Is a cumulative refund of `total` over the committed amount? Derived from the
 * canonical I-1 by probing {@link checkI1ValueConservation} with a commitment in
 * `Refunded(total)` state — so this is the point-in-time rule applied to the sum,
 * not a second copy of it.
 */
function isCumulativeOverRefund(order: Commitment, total: number, currency: Money["currency"]): boolean {
  const probe: Commitment = {
    ...order,
    state: { type: "Refunded", amount: { amount: total, currency }, at: order.created_at },
    history: [],
  };
  return checkI1ValueConservation([probe]).some(
    (v) => v.invariant === "I-1" && v.description.includes("cannot exceed what was captured"),
  );
}

/**
 * The identity of an action for replay detection. Two actions are "the same" iff
 * this key is equal. An explicit `idempotencyKey` is used when supplied (the
 * unambiguous, standard approach); otherwise a fingerprint is derived from the
 * fields that define the operation — commitment, target type, amount (for a
 * Refunded move), and actor. Two genuinely-distinct but structurally-identical
 * actions therefore need distinct keys to be applied separately.
 */
function actionKey(action: ProposedAction): string {
  if (action.idempotencyKey !== undefined) return `key:${action.idempotencyKey}`;
  const parts = [action.commitment, action.to.type, String(action.actor)];
  if (action.to.type === "Refunded") {
    parts.push(String(action.to.amount.amount), action.to.amount.currency);
  }
  return `fp:${parts.join("|")}`;
}

/**
 * The root of `commitment`'s tree in `world` — walk `parent` pointers up while the
 * parent is present in the world. A commitment with no parent (or whose parent is
 * not in the world) is its own root. The append-only `parent` / `children` fields
 * already exist on the model, so this is a pure structural read — no schema change.
 */
function treeRootOf(world: World, commitment: Commitment): Commitment {
  const byId = new Map(world.commitments.map((c) => [c.id as string, c]));
  let current = commitment;
  const seen = new Set<string>();
  while (current.parent !== undefined) {
    const parent = byId.get(current.parent as string);
    if (parent === undefined || seen.has(current.id as string)) break;
    seen.add(current.id as string);
    current = parent;
  }
  return current;
}

/** A commitment is "in a tree" if it has a parent (is a child) or has children. */
function isInTree(commitment: Commitment, root: Commitment): boolean {
  return (root.id as string) !== (commitment.id as string) || commitment.children.length > 0;
}

export function createSession(initialWorld: World): Session {
  let world = initialWorld;
  const ledger = new Map<string, RefundTally>();
  // Per-TREE cumulative refund ledger, keyed by the tree ROOT id. The per-commitment
  // `ledger` above caps each commitment against its own committed amount; this caps
  // the SUM of refunds across a parent + its children against the parent's committed
  // amount (full multi-object coherence). Standalone commitments are never tree
  // members, so this leaves single-commitment behaviour byte-for-byte unchanged.
  const treeLedger = new Map<string, RefundTally>();
  // Idempotency / replay-safety: keys of actions ALREADY APPLIED in this session.
  // Scope is per-session and in-memory — durable, cross-session idempotency would
  // need a persistent store and is not provided here (see the module docs).
  const applied = new Set<string>();

  function propose(action: ProposedAction): GuardResult {
    // Replay detection: if this exact action was already applied in this session,
    // it is a no-op — do NOT apply it again, do NOT advance the world. Report the
    // current (unchanged) world flagged `replay: true` so the caller learns
    // "already done", never "applied twice". Only ACCEPTED actions are recorded,
    // so re-proposing a previously-rejected action is re-evaluated normally.
    const key = actionKey(action);
    if (applied.has(key)) {
      return { ok: true, next: world, replay: true };
    }

    // Optimistic-concurrency conflict (distinct from a replay): if this action was
    // planned against a version the commitment has since moved past (a concurrent
    // actor advanced it in this accumulated world), reject as a CONFLICT so the
    // caller re-reads. Checked here too because the partial-refund path below does
    // not route through guardAction. A replay (handled above) is NOT a conflict.
    const conflictTarget = world.commitments.find((c) => (c.id as string) === action.commitment);
    if (conflictTarget !== undefined) {
      const conflict = checkVersion(conflictTarget, action.expectedVersion);
      if (conflict) return conflict;
    }

    // Refund actions get the cross-step cumulative check; everything else is a
    // straight compose over guardAction.
    if (action.to.type === "Refunded") {
      const order = world.commitments.find((c) => (c.id as string) === action.commitment);
      // If the order can't legally reach Refunded from its current state (e.g. a
      // refund proposed before the order was ever fulfilled, or after it is fully
      // refunded), let guardAction produce the I-2 rejection WITH alternatives.
      if (order === undefined || !isValidCommitmentTransition(order.state, action.to)) {
        return guardAction(world, action);
      }

      const committed = committedTotal(order);
      const proposed = action.to.amount;
      if (committed !== null && proposed.currency === committed.currency) {
        const prior = ledger.get(action.commitment);
        const priorAmt = prior ? prior.total.amount : 0;
        const priorCount = prior ? prior.count : 0;
        const cumulative = priorAmt + proposed.amount;

        if (isCumulativeOverRefund(order, cumulative, committed.currency)) {
          const remaining = Math.max(0, committed.amount - priorAmt);
          return {
            ok: false,
            violations: [
              {
                rule: "I-1",
                message:
                  `Cumulative refunds on ${order.id} would reach ${cumulative} ${committed.currency} across ` +
                  `${priorCount + 1} refund(s), but only ${committed.amount} ${committed.currency} was committed — ` +
                  `value is not conserved across the session (the point-in-time check sees each refund alone).`,
                fix:
                  `Refund at most the remaining ${remaining} ${committed.currency} ` +
                  `(committed ${committed.amount} − already refunded ${priorAmt}).`,
              },
            ],
            alternatives: [
              {
                to: "Refunded",
                label: "refund the commitment",
                bounded: `cumulative refunds must stay within the committed ${committed.amount} ${committed.currency}; ${remaining} ${committed.currency} remains refundable`,
              },
            ],
          };
        }

        // Multi-object coherence: if this commitment is part of a tree (a parent
        // with children, or a child), cap the SUM of refunds across the whole tree
        // against the PARENT's committed amount — so refunds spread over different
        // children (each individually valid, each child reconciling via I-6) cannot
        // cumulatively exceed the parent. Composes the existing checkI6TreeConsistency
        // (structure) + the same I-1 cumulative probe (lifted to the parent).
        const root = treeRootOf(world, order);
        const treeMember = isInTree(order, root);
        if (treeMember) {
          const children = world.commitments.filter((c) => c.parent === root.id);
          const i6 = checkI6TreeConsistency(root, children);
          if (i6.length > 0) {
            return { ok: false, violations: i6.map((v) => ({ rule: v.invariant, message: v.description, fix: v.fix })) };
          }
          const treeCommitted = committedTotal(root);
          if (treeCommitted !== null && proposed.currency === treeCommitted.currency) {
            const treePrior = treeLedger.get(root.id as string);
            const treePriorAmt = treePrior ? treePrior.total.amount : 0;
            const treeCount = treePrior ? treePrior.count : 0;
            const treeCumulative = treePriorAmt + proposed.amount;
            if (isCumulativeOverRefund(root, treeCumulative, treeCommitted.currency)) {
              const remaining = Math.max(0, treeCommitted.amount - treePriorAmt);
              return {
                ok: false,
                violations: [
                  {
                    rule: "I-1",
                    message:
                      `Cumulative refunds across the commitment tree rooted at ${root.id} would reach ` +
                      `${treeCumulative} ${treeCommitted.currency} across ${treeCount + 1} refund(s) on the ` +
                      `parent and its children, but the parent committed only ${treeCommitted.amount} ` +
                      `${treeCommitted.currency} — value is not conserved across the tree.`,
                    fix:
                      `Refund at most the remaining ${remaining} ${treeCommitted.currency} across the tree ` +
                      `(parent committed ${treeCommitted.amount} − already refunded across the tree ${treePriorAmt}).`,
                  },
                ],
                alternatives: [
                  {
                    to: "Refunded",
                    label: "refund the commitment",
                    bounded: `cumulative refunds across the tree must stay within the parent's committed ${treeCommitted.amount} ${treeCommitted.currency}; ${remaining} ${treeCommitted.currency} remains refundable`,
                  },
                ],
              };
            }
          }
        }

        // Accepted refund. Record it in the ledger. Keep the order in Fulfilled
        // for a PARTIAL refund (the schema has no partial-refund state); transition
        // it to Refunded only once the refunds reach the committed total.
        const newTotal = prior ? add(prior.total, proposed) : proposed;
        const recordTree = () => {
          if (!treeMember) return;
          const tp = treeLedger.get(root.id as string);
          treeLedger.set(root.id as string, {
            total: tp ? add(tp.total, proposed) : proposed,
            count: (tp ? tp.count : 0) + 1,
          });
        };
        const fullyRefunded = moneyEquals(cumulative, committed.amount, committed.currency);
        if (fullyRefunded) {
          // A real Fulfilled → Refunded transition for the final, full refund.
          const verdict = guardAction(world, action);
          if (!verdict.ok) return verdict;
          world = verdict.next;
          ledger.set(action.commitment, { total: newTotal, count: priorCount + 1 });
          recordTree();
          applied.add(key);
          return verdict;
        }
        ledger.set(action.commitment, { total: newTotal, count: priorCount + 1 });
        recordTree();
        applied.add(key);
        return { ok: true, next: world };
      }
    }

    // Non-refund action: pure compose over guardAction.
    const verdict = guardAction(world, action);
    if (verdict.ok) {
      world = verdict.next;
      applied.add(key);
    }
    return verdict;
  }

  return {
    propose,
    get world() {
      return world;
    },
    refundedSoFar(commitmentId: string): Money | null {
      const tally = ledger.get(commitmentId);
      return tally ? tally.total : null;
    },
  };
}
